Mathlib/Tactic/Congr!.lean

```diff
@@ -0,0 +1,248 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean
+import Mathlib.Tactic.Relation.Rfl
+
+/-!
+# The `congr!` tactic
+
+This is a more powerful version of the `congr` tactic.
+-/
+
+open Lean Meta Elab Tactic
+
+/--
+Asserts the given congruence theorem as fresh hypothesis, and then applies it.
+Return the `fvarId` for the new hypothesis and the new subgoals.
+-/
+private def applyCongrThm? (mvarId : MVarId) (congrThmType congrThmProof : Expr) :
+    MetaM (List MVarId) := do
+  let mvarId ← mvarId.assert (← mkFreshUserName `h_congr_thm) congrThmType congrThmProof
+  let (fvarId, mvarId) ← mvarId.intro1P
+  let mvarIds ← mvarId.apply (mkFVar fvarId) { synthAssignedInstances := false }
+  mvarIds.mapM fun mvarId => mvarId.tryClear fvarId
+
+/--
+Try applying user-provided congruence lemmas. If any are applicable,
+returns a list of new goals.
+-/
+def Lean.MVarId.userCongr? (mvarId : MVarId) : MetaM (Option (List MVarId)) :=
+  mvarId.withContext do
+    mvarId.checkNotAssigned `userCongr?
+    let some (lhs, _) := (← mvarId.getType').eqOrIff? | return none
+    let some name := lhs.getAppFn.cleanupAnnotations.constName? | return none
+    let congrTheorems := (← getSimpCongrTheorems).get name
+    -- Note: congruence theorems are in provided in decreasing order of priority.
+    for congrTheorem in congrTheorems do
+      let res ← observing? do
+        let cinfo ← getConstInfo congrTheorem.theoremName
+        let us ← cinfo.levelParams.mapM fun _ => mkFreshLevelMVar
+        let proof := mkConst congrTheorem.theoremName us
+        let ptype ← instantiateTypeLevelParams cinfo us
+        applyCongrThm? mvarId ptype proof
+      if let some mvars := res then
+        return mvars
+    return none
+
+/--
+Try to convert an `Iff` into an `Eq` by applying `iff_of_eq`.
+If successful, returns the new goal, and otherwise returns the original `MVarId`.
+-/
+def Lean.MVarId.iffOfEq (mvarId : MVarId) : MetaM MVarId := do
```

Comment:
Is this a special case of `liftReflToEq`?

Reply:
Yeah, it's that but specifically for `iff`, generating a more straightforward proof. Should I do anything more than add a comment about this?

Reply:
No, if the generated proof is easier, it is probably nice to keep it separate.